osbrain/tests/test_agent_async_requests.py

```diff
@@ -22,114 +23,113 @@ def on_error(agent):
-
-    # Client requests should be non-blocking
-    t0 = time.time()
-    client.send('async', 'foo')
-    client.send('async', 'bar')
-    assert time.time() - t0 < 0.1
-
-    # Server should receive first request "soon"
-    time.sleep(0.1)
-    assert server.get_attr('received') == ['foo']
+    assert wait_agent_attr(server1, value=['foo'], timeout=.1)
+    assert wait_agent_attr(server2, value=['bar'], timeout=.1)
     assert client.get_attr('received') == []
 
-    # Wait for client to receive reply asynchronously
-    time.sleep(1)
-    assert server.get_attr('received') == ['foo', 'bar']
-    assert client.get_attr('received') == ['xfoo']
-    time.sleep(1)
-    assert server.get_attr('received') == ['foo', 'bar']
-    assert client.get_attr('received') == ['xfoo', 'xbar']
+    server2.unsafe.set_attr(blocked=False)
+    assert wait_agent_attr(client, value=['xbar'], timeout=.1)
+
+    server1.unsafe.set_attr(blocked=False)
+    assert wait_agent_attr(client, value=['xbar', 'xfoo'], timeout=.1)
 
 
 def test_yield(nsproxy):
     """
     Asynchronous request-reply pattern with a reply handler that yields.
     """
-    def late_reply(agent, request):
+    def blocked_reply_yield(agent, request):
         agent.received.append(request)
-        time.sleep(1)
+        agent.blocked = True
+        while agent.blocked:
+            time.sleep(.01)
+
         yield 'x' + request
         agent.received.append('y' + request)
 
-    server = run_agent('server', base=Server)
-    client = run_agent('client', base=Client)
-
-    addr = server.bind('ASYNC_REP', alias='replier', handler=late_reply)
-    client.connect(addr, alias='async', handler=append_received)
-
-    # Client requests should be non-blocking
-    t0 = time.time()
-    client.send('async', 'foo')
-    client.send('async', 'bar')
-    assert time.time() - t0 < 0.1
+    server1, server2, client = async_client_servers(blocked_reply_yield)
+    client.send('async1', 'foo')
+    client.send('async2', 'bar')
 
-    # Server should receive first request "soon"
-    time.sleep(0.1)
-    assert server.get_attr('received') == ['foo']
+    assert wait_agent_attr(server1, value=['foo'], timeout=.1)
+    assert wait_agent_attr(server2, value=['bar'], timeout=.1)
     assert client.get_attr('received') == []
 
-    # Wait for client to receive reply asynchronously
-    time.sleep(1)
-    assert server.get_attr('received') == ['foo', 'yfoo', 'bar']
-    assert client.get_attr('received') == ['xfoo']
-    time.sleep(1)
-    assert server.get_attr('received') == ['foo', 'yfoo', 'bar', 'ybar']
-    assert client.get_attr('received') == ['xfoo', 'xbar']
+    server2.unsafe.set_attr(blocked=False)
+    assert wait_agent_attr(client, value=['xbar'], timeout=.1)
+    assert server2.get_attr('received') == ['bar', 'ybar']
+
+    server1.unsafe.set_attr(blocked=False)
+    assert wait_agent_attr(client, value=['xbar', 'xfoo'], timeout=.1)
+    assert server1.get_attr('received') == ['foo', 'yfoo']
```

Comment:
Use `wait_agent_attr()` instead of `server.get_attr('received')`. For consistency and also to avoid (very unlikely) race conditions.